Feature: #45 Modal.

damontgomery opened the pull request:
# Description

Adds a modal. See #45 

# Testing notes
- Visit http://localhost:6006/?path=/story/molecules-modal--small and similar examples.

# Example

![modal_canvas](https://user-images.githubusercontent.com/397902/125838470-3af59c37-6506-4501-bb28-6b8a0bac365f.jpg)

## Documentation

![modal_documentation](https://user-images.githubusercontent.com/397902/125838376-f882a503-bbda-4456-9caa-2e6c19bab94c.jpg)

# Notes

Took concepts from https://micromodal.vercel.app and https://patternflyelements.org/components/modal

The controls for slots in Storybook are weird and this documentation works, but is awkward. Open to suggestions!


<a href="https://gitpod.io/#https://github.com/phase2/outline/pull/49"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

damontgomery committed:
45: Modal story and basic component.

damontgomery committed:
45: Add default size for modal.

damontgomery committed:
45: Update stories for modal.

- Use outline components which are styled.
- Show use of slots.

damontgomery committed:
45: Add modal open / close behavior.

damontgomery committed:
45: Styles and keyboard navigation.

damontgomery committed:
45: Prettier formatting updates.

damontgomery commented on src/components/base/outline-modal/outline-modal.stories.ts:
```diff
@@ -0,0 +1,106 @@
+}): TemplateResult => {
+  return html`
+    <outline-modal size="${ifDefined(size)}">
+      ${ifDefined(triggerSlot)} ${ifDefined(headerSlot)}
```
Prettier?? Not sure why it doesn't want this on a separate line. :(

damontgomery committed:
45: Add events to the modal.

damontgomery committed:
45: Escape to close and focus.

damontgomery committed:
45: Set modal z-index so it overlays other things.

damontgomery committed:
45: Update documentation with better code example and grouped arguments.

damontgomery committed:
45: Add documentation around the modal.

damontgomery committed:
45: Tailwind styles.

damontgomery commented:
I worked on documentation notes to describe functionality and usage.

Added events so others can detect what's happening and respond to changes.

Switched to using Tailwind classes in most cases.

Still need to add styles to different sizes of the modal (small, medium, full-screen).

damontgomery committed:
45: Fix Tailwind styles for customizations.

Tailwind styles weren't actually applying because of compile issues.

damontgomery committed:
45: Add different sizes for the modal.

damontgomery commented:
I added the styles for different sizes. Going to review this later today and see if there are other things I need to add.

damontgomery committed:
45: Add documentation to the custom element.

Remove the comment for `size` since it doesn't provide any extra information.

damontgomery commented:
Going to review the W3 requirements, https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/dialog.html.

damontgomery committed:
45: Prettier formatting on the story.

:( Why would we be able to use spacing in the template how we like?

damontgomery committed:
45: Accessibility: full screen on small screens.

damontgomery committed:
45: Accessibility references for the modal if available.

- aria-describedby
- aria-labelledby

LeeMellon commented on src/components/base/outline-modal/outline-modal.css:
```diff
@@ -0,0 +1,40 @@
+* {
+  @apply box-border;
```
Not 100% sure this is needed, but I could be wrong.

damontgomery replied:
Thanks for taking a look!

I don't know either, but I saw it in a best-practices. Maybe it's part of the global styles? I can try removing.

damontgomery replied:
It didn't seem to do anything here, so I removed it. Thanks again.

LeeMellon replied:
Yeah I think that is part of global styles.

LeeMellon commented on src/components/base/outline-modal/outline-modal.ts:
```diff
@@ -0,0 +1,134 @@
+          id="overlay"
+          tabindex="-1"
+          class="${this.size}"
+          @click="${this._handleModalClose}"
```
Wondering if it might be worth it to make this conditional? In the case where you may want to force the user to take a specific action before closing the modal, rather than just being able to click the overlay to close it.

damontgomery replied:
It's a good question. I don't have strong opinions. Maybe we can discuss as a group?

himerus replied:
This would be a nice feature. Another prop of `canClose` or something, but this would require additional work too so we define what the confirmation button/link/text is/does/etc. 

If it's not "easy", I'd say we add this as a new feature request and iterate.

damontgomery replied:
Thanks for taking a look. https://designsystem.digital.gov/components/modal/ has an example of this we can use as a starting point. For now, I think making the close button optional and requiring the developer to configure the actions is easiest / most flexible and we could look at other options later?

I can probably take a look today.

damontgomery replied:
I've added this as an attribute `shouldForceAction`. There is a story for this in the latest version. Should remove the close features and adjusts the keyboard navigation for focus.

damontgomery committed:
45: Return focus to trigger when modal is closed.

damontgomery committed:
45: Modal documentation refactor.

Try to put more broadly useful information first.

LeeMellon reviewed (approved):
Looks fantastic! Two comments that probably speak more to my ignorance than anything. Would still be okay merging this.

damontgomery committed:
45: Focus on the first focusable element when opening the modal.

damontgomery commented:
I tried to add a bunch of accessibility features around focus. Still need to trap the focus within the modal.

It also surfaced an issue where I wasn't able to focus our `outline-button` element. Not sure why.

damontgomery committed:
45: Update documentation for modal.

- update code sample
- add notes on which examples are accessibility features
- add story for accessibility description

damontgomery committed:
45: Refactor the component for clarity.

Try to group things logically.

damontgomery committed:
45: Fix bug on Docs page where keyboard events weren't registering.

damontgomery committed:
45: Trap focus within the modal.

damontgomery committed:
45: Add more detailed button example.

damontgomery committed:
45: Remove border box since it doesn't seem to do anything here.

LeeMellon commented on src/components/base/outline-modal/outline-modal.ts:
```diff
@@ -0,0 +1,263 @@
+
+// See https://stackoverflow.com/questions/1599660/which-html-elements-can-receive-focus.
+// @todo make this re-usable across components?
+const focusableElementSelector = `
```
I don't know if you need this. This is describing default behaviors, and in the case of the modal I think you can provide prop that takes an element id that gets focus when the modal opens, and that defaults to the "#main" div wrapping your default slot.

damontgomery commented on src/components/base/outline-modal/outline-modal.ts:
```diff
@@ -0,0 +1,263 @@
+  @query('#trigger')
+  private triggerElement!: HTMLDivElement;
+
+  private _handleModalTrigger(event: MouseEvent | KeyboardEvent): void {
+    let shouldOpen = false;
+
+    switch (event.type) {
+      case 'click':
+        shouldOpen = true;
+        break;
+      case 'keydown':
+        if ('key' in event && event.key === 'Enter') {
+          shouldOpen = true;
+          // This prevents a focused element from also triggering.
+          // For example, the modal opens and the "accept" button is focused and then triggered and the modal closes.
+          event.preventDefault();
+          break;
+        }
+    }
```
I think I should probably split this into two handlers. One for the MouseEvent and one for the KeyboardEvent.

The code is a bit difficult to read as-is, because it takes two different inputs, spends most of the time figuring out which of those we are looking at, and then does something very simple.

I'll try to give this a shot on Monday.

damontgomery commented on src/components/base/outline-modal/outline-modal.ts:
```diff
@@ -0,0 +1,263 @@
+          tabindex="-1"
+          class="${this.size}"
+          @click="${this._handleModalClose}"
+          @keydown="${this._handleOverlayKeyup}"
```
This needs to be updated. When I switched from `keyup` to `keydown`, I forgot to rename the method. :(

I'll try to tackle that on Monday.

damontgomery commented on src/components/base/outline-modal/outline-modal.ts:
```diff
@@ -0,0 +1,263 @@
+                id="close"
+                aria-label="Close modal"
+                @click="${this._handleModalClose}"
+                @keydown="${this._handleModalCloseKeyup}"
```
And mismatch between `keyup` and `keydown` here too. :(

damontgomery committed:
45: Clean up code.

- Remove confusing switch statements
- Name handlers to match events
- Remove implicit "modal" from methods

damontgomery committed:
45: Remove handlers that just call another method.

These can be re-created if the code becomes more complex.

damontgomery commented on src/components/base/outline-modal/outline-modal.ts:
```diff
@@ -0,0 +1,251 @@
+  private _trapFocus(): void {
+    // We will use the close button as the first focusable element.
+
+    let lastFocusableElement: HTMLElement = this.closeElement;
+
+    const focusableElements: NodeListOf<HTMLElement> = this.querySelectorAll(
+      focusableElementSelector
+    );
+
+    if (focusableElements.length > 0) {
+      lastFocusableElement = focusableElements[focusableElements.length - 1];
+    }
```
Would this make more sense?

```typescript
const focusableElements: NodeListOf<HTMLElement> = this.querySelectorAll(focusableElementSelector);

const lastFocusableElement: HTMLElement = focusableElements[focusableElements.length - 1] ?? this.closeElement;
```

This avoids re-assigning to `lastFocusableElement`.

damontgomery commented on src/components/base/outline-modal/outline-modal.ts:
```diff
@@ -0,0 +1,251 @@
+  elementToFocusSelector?: string | undefined;
+
+  private _focusOnElement(): void {
+    let elementToFocus: HTMLElement = this.closeElement;
+
+    if (this.elementToFocusSelector !== undefined) {
+      const attributeDefinedElementToFocus = this.querySelector(
+        this.elementToFocusSelector
+      ) as HTMLElement | null;
+
+      if (attributeDefinedElementToFocus !== null) {
+        elementToFocus = attributeDefinedElementToFocus;
+      }
+    }
+
+    if (this.elementToFocusSelector === undefined) {
+      const automaticallySelectedElementToFocus = this.querySelector(
+        focusableElementSelector
+      ) as HTMLElement | null;
+
+      if (automaticallySelectedElementToFocus !== null) {
+        elementToFocus = automaticallySelectedElementToFocus;
+      }
+    }
+
+    elementToFocus.focus();
```
Does this make more sense? It avoids changing a variables values.

```typescript
const defaultElement: HTMLElement = this.closeElement;

const attributeDefinedElement = (this.elementToFocusSelector !== undefined) ? this.querySelector(this.elementToFocusSelector) as HTMLElement | null : null;

const automaticallySelectedElement = this.querySelector(focusableElementSelector) as HTMLElement | null;

const element = attributeDefinedElement ?? automaticallySelectedElement ?? defaultElement;

element.focus();
```

damontgomery committed:
45: Replace simple null checks with ternary.

This seems to make it easier to read by putting the label / conditional next to the attribute.

damontgomery committed:
45: Refactor focus on element to avoid changing the value of variables.

damontgomery committed:
45: More descriptive variable names and avoid changing value on trap focus.

damontgomery committed:
45: Use small get method to make trap focus clearer.

damontgomery committed:
45: Use slot state.

damontgomery committed:
Merge branch 'next' into feature/45-modal

damontgomery committed:
Merge branch 'next' into feature/45-modal

damontgomery committed:
test: 45: Outline modal test.

Provide an interface so the test knows what properties and methods are available.

Test the slots are filled as expected when using the modal.

damontgomery committed:
chore: 45: Use generic name for Outline Component to help with copy & paste.

damontgomery committed:
feature: Story for force action on modal.

damontgomery committed:
Feature: 45: Force action feature for modal.

damontgomery commented:
Added an automated test that checks that the slots are filled when the modal is rendered, opens, and closes.

Also added the `shouldForceAction` attribute so we can force the user to take an action if needed.

damontgomery committed:
chore: 45: Reduce copy & paste in type definition for modal.

mabry1985 commented on src/components/base/outline-modal/outline-modal.ts:
```diff
@@ -0,0 +1,282 @@
+                name="outline-modal--header"
+                @slotchange="${this._handleSlotChange}"
+              ></slot>
+              ${this.shouldForceAction
```
Can we place `this.shouldForceAction` in it's own template function? See [here](https://github.com/phase2/outline/pull/60#discussion_r679571323) for examples

mabry1985 commented on src/components/base/outline-modal/outline-modal.ts:
```diff
@@ -0,0 +1,282 @@
+            id="container"
+            role="dialog"
+            aria-modal="true"
+            aria-labelledby="${ifDefined(
```
Let's move this conditional logic for `aria-labelledby` and `aria-described` into variables at the beginning of the render function to clean up this template a bit:
```
render() {
const ariaDesc = this._hasAccessibilityDescriptionSlot ? 'accessibility-description' : null
const ariaLabel = this._hasHeaderSlot ? 'header' : null
 ....
 <div aria-labeledby="${ifDefined(ariaLabel)} aria-describedby="${ifDefined(ariaDesc}">
....
}
```

mabry1985 reviewed (changes requested):
I think this is looking pretty good. I appreciate all the great accessibility work here!

I've suggested a couple changes to refactor a bit. 

I would also like to see some more styling done on this component, but lets make an issue for that and I can address that in a separate PR

mabry1985 reviewed (approved):
I think this is looking good, there were a couple changes that were discussed previously that I resolved based off of our meeting earlier and I think this is ready to come in and be iterated on.
